<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Tutorial written by Kevin Burfeind" />
<meta name="author" content="Tutorial revised by Jonathan Nelson" />
<meta name="author" content="Tutorial adapted by Sienna Blanche" />

<meta name="date" content="2025-10-21" />

<title>Lesson 2</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Nelson Lab</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Lab Info
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="lab_schedule.html">Lab Schedule</a>
    </li>
    <li>
      <a href="lab_members.html">Lab Members</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Getting Started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="computer_requirements.html">Computer Requirements</a>
    </li>
    <li>
      <a href="downloading_R.html">Downloading R</a>
    </li>
    <li>
      <a href="using_packages.html">Using Packages</a>
    </li>
    <li>
      <a href="best_practices.html">Jonathan's Best Coding Practices</a>
    </li>
    <li>
      <a href="rules_of_coding.html">Jonathan's Rule of Coding</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Github Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="qupath_install.html">Understanding Github</a>
    </li>
    <li>
      <a href="github_tutorial.html">Github Tutorial</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    snRNAseq Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="lesson_1.html">Lesson 1</a>
    </li>
    <li>
      <a href="lesson_2.html">Lesson 2</a>
    </li>
    <li>
      <a href="write_degs_excel.html">Write DEGs to Excel</a>
    </li>
    <li>
      <a href="coding_crumbs.html">SCORE: Coding Crumbs</a>
    </li>
    <li>
      <a href="recreating_figures.html">SCORE: Recreating Figures</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    QuPath Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="qupath_install.html">Installing QuPath</a>
    </li>
    <li>
      <a href="qupath_manual_annotation.html">Manual Cell Annotation</a>
    </li>
    <li>
      <a href="qupath_composite_classifications.html">Creating Composite Classifications</a>
    </li>
    <li>
      <a href="qupath_analyze_in_r.html">Analyzing QuPath data in R</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Problem Solving
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="write_degs_excel.html">Write DEGs to Excel</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Lesson 2</h1>
<h4 class="author">Tutorial written by Kevin Burfeind</h4>
<h4 class="author">Tutorial revised by Jonathan Nelson</h4>
<h4 class="author">Tutorial adapted by Sienna Blanche</h4>
<h4 class="date">21 October 2025</h4>

</div>


<p>At this point everyone should have RStudio installed along with
Seurat. Please refer to the <a
href="https://usckrc.github.io/Education_Website/downloading_R.html">Downloading
R</a> section of the website if you need to do that again.</p>
<div id="getting-started" class="section level1">
<h1>Getting Started</h1>
<p>Create the “snRNA_seq_learning” folder on your desktop if you have
not already done so. Properly set your working directory to the
<strong>snRNA_seq_learning folder</strong> by clicking on
<strong>Session&gt;Set Working Directory&gt;Choose Directory</strong>.
This tells R studio where our downloaded files are.</p>
<p><img
src="images/lesson_1/ROC_1_Screenshot%200.png" /></p>
<p>When the file browser opens, navigate to and <strong>select the
scRNAseq_learning folder</strong>, then select open. This sets the
working directory so RStudio knows where to look for your files.</p>
<p><img
src="images/lesson_1/ROC_1_Screenshot%201.1.png" /></p>
<div id="necessary-materials" class="section level2">
<h2>Necessary Materials</h2>
<p>You can download the necessary file from the <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE111107">GEO
entry</a> and click the sample titled
<strong>“Glom_rep1”(GSM3022239)</strong> then download it by clicking on
the “ftp” link next to <strong>GSM3022239_dge_glom_rep1.txt.gz</strong>
at the bottom of the page.</p>
<p><img
src="images/lesson_2/roc_2_screenshot%201.png" /></p>
<p><img
src="images/lesson_2/roc_2_screenshot%202.png" /></p>
<p>Drag and drop the file into the “snRNA_seq_learning” folder. Double
click on the file to unzip it. Do the same for the sample titled
<strong>“Glom_rep2” (GSM3022240)</strong>.</p>
<p>Alternatively, you can download it from [Google Drive] NEED (<a
href="https://drive.google.com/file/d/16JzkZN2qmUzK8bg1_D2x3vRstWrqvHUE/view?usp=sharing"
class="uri">https://drive.google.com/file/d/16JzkZN2qmUzK8bg1_D2x3vRstWrqvHUE/view?usp=sharing</a>)
(<strong>recommended</strong>). To follow along with this tutorial you
can download the [R Script] NEED (<a
href="https://drive.google.com/file/d/1RODN4wE31qgnaK6_WjnC1BihtVgX7fMu/view?usp=sharing"
class="uri">https://drive.google.com/file/d/1RODN4wE31qgnaK6_WjnC1BihtVgX7fMu/view?usp=sharing</a>)
and run it or create your own and copy the code from the tutorial into
the script.</p>
</div>
</div>
<div id="step-1" class="section level1">
<h1>Step 1</h1>
<div id="activate-the-r-packages-needed-to-run-the-analysis."
class="section level2">
<h2>Activate the R packages needed to run the analysis.</h2>
<p>Copy the following lines of code into your file and run them:</p>
<p><strong>library(Seurat)</strong></p>
<p><strong>library(dplyr)</strong></p>
<p><strong>library(patchwork)</strong></p>
</div>
</div>
<div id="step-2" class="section level1">
<h1>Step 2</h1>
<div id="upload-the-data" class="section level2">
<h2>Upload the data</h2>
<p>We will be uploading the data slightly differently this time, by
using the read.table function. Copy and paste the following code which
converts the .txt file into a data file.</p>
<p><strong>Glom_rep1.data &lt;- read.table(file =
“snRNAseq_learning/GSM3022239_deg_glom_rep1.txt”, header = TRUE,
row.names = 1, colClasses =c(“character”, rep(“numeric”)))</strong></p>
<p>Glom_rep1.data is a data frame that can be utilized in several
different ways. We will next create a seurat object out of the data
frame using the following code.</p>
<p><strong>Glom_rep1 &lt;- CreateSeuratObject(counts = Glom_rep1.data,
project = “Glom_merged”, min.cells = 3, min.features = 200)</strong></p>
<p><strong>Glom_rep1</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%203.png" /></p>
<p>Now Glom_rep1 is an object of class Seurat with 17511 features
(genes) across 4556 samples (cells). Since we are planning to merge this
dataset with another, we will need to check to see if its identity is
coded in the metadata.</p>
<p><strong>head(<a href="mailto:Glom_rep1@meta.data"
class="email">Glom_rep1@meta.data</a>)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%204.png" /></p>
<p>As you can see, the column orig.ident contains “Glom_merged”, which
is is the project name. Once we add another dataset, that will be
overwritten, and can’t be used to pull out data. We will add a column to
the metadata that denotes the replicate.</p>
<p><strong>Glom_rep1 &lt;- AddMetaData(object = Glom_rep1, metadata =
“rep1”, col.name = “replicate”)</strong></p>
<p><strong>head(<a href="mailto:Glom_rep1@meta.data"
class="email">Glom_rep1@meta.data</a>)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%205.png" /></p>
<p>Next, make the second second replicate a data frame.</p>
<p><strong>Glom_rep2.data &lt;- read.table(file =
“Glom/GSM3022240_dge_glom_rep2.txt”, header = TRUE, row.names = 1,
colClasses =c(“character”, rep(“numeric”)))</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%206.png" /></p>
<p>We will now create a seurat object out of Glom_rep2 as we did for
Glom_rep1.</p>
<p><strong>Glom_rep2 &lt;- CreateSeuratObject(counts = Glom_rep2.data,
project = “Glom_merged”, min.cells = 3, min.features = 200)</strong></p>
<p><strong>Glom_rep2</strong></p>
<p>Now Glom_rep2 is an object of class Seurat with 18309 features
(genes) across 4366 samples (cells). We will next add a column to the
metadata indicate the replicate.</p>
<p><img
src="images/lesson_2/roc_2_screenshot%207.png" /></p>
<p><strong>Glom_rep2 &lt;- AddMetaData(object = Glom_rep2, metadata =
“rep2”, col.name = “replicate”)</strong></p>
<p><strong>head(<a href="mailto:Glom_rep2@meta.data"
class="email">Glom_rep2@meta.data</a>)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%208.png" /></p>
</div>
</div>
<div id="step-3" class="section level1">
<h1>Step 3:</h1>
<div id="merge-the-two-seurat-objects." class="section level2">
<h2>Merge the two Seurat objects.</h2>
<p>To merge the two objects, we will use the “merge” function, but there
are several different ways to merge Seurat objects. Note: the “merge”
function does not work when datasets are very different from one another
- i.e., stim cell vs. unstim cell. However, It does seem to work when
comparing disease vs. no disease</p>
<p><strong>Glom_merged &lt;- merge(Glom_rep1, y = Glom_rep2,
add.cell.ids = c(“Glom_rep1”, “Glom_rep2”), project =
“Glom_merged”)</strong></p>
<p><strong>Glom_merged</strong></p>
<p>Glom_merged is now an object with 19183 features (genes) across 8922
samples (cells). Note that the sample number is the sum of the two reps:
4556 + 4366 = 8922.</p>
<p><img
src="images/lesson_2/roc_2_screenshot%209.png" /></p>
<p>Also notice the cell names now have an added identifier</p>
<p><strong>head(colnames(Glom_merged))</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2010.png" /></p>
<p>We can also double check the number of cells in each group using the
“table” function, pulling out “replicate”.</p>
<p><strong>table(Glom_merged$replicate)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2011.png" /></p>
<p>Now, begin the QC that we did last lesson, starting with filtering
out samples with a high % of mito genes. This adds a column in the
metadata called “percent.mito”.</p>
<p><strong>Glom_merged[[“percent.mt”]] &lt;-
PercentageFeatureSet(Glom_merged, pattern = “^mt-”)</strong></p>
<p><strong>head(<a href="mailto:Glom_merged@meta.data"
class="email">Glom_merged@meta.data</a>)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2012.png" /></p>
<p>As with last lesson, visualize the dataset for quality control using
3 different metrics:</p>
<p>nFeature_RNA = Number of differing genes detected in each cell
nCount_RNA = Number of mRNA molecules detected in each cell percent.mt =
Percent of counts that come from mitochondrial genes</p>
<p><strong>VlnPlot(Glom_merged, features = c(“nFeature_RNA”,
“nCount_RNA”, “percent.mt”), ncol = 3)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2013.png" /></p>
<p>We will interpret this information to filter out doublets and damaged
cells. High nFeature_RNA or nCount_RNA could equate to doublets or mixed
cell debris. We will filter out cells with more than 2000 genes and 4000
mRNA molecules. High percent.mt could equate to damaged cells. We will
filter out cells with more than 20% mitochondrial content.</p>
<p><strong>Glom_merged &lt;- subset(Glom_merged, subset = nFeature_RNA
&lt; 2000 &amp; nFeature_RNA &lt; 4000 &amp; percent.mt &lt;
20)</strong></p>
<p>Now we will replot the graphs to visualize the plots again with the
newly filtered data</p>
<p><strong>VlnPlot(Glom_merged, features = c(“nFeature_RNA”,
“nCount_RNA”, “percent.mt”), ncol = 3)</strong></p>
<p><strong>Glom_merged</strong></p>
<p>Glom_merged is now an object with 19183 features (genes) across 8361
samples (cells). This filtered out 561 cells and 0 genes!</p>
<p><img
src="images/lesson_2/roc_2_screenshot%2014.png" /></p>
</div>
</div>
<div id="step-4" class="section level1">
<h1>Step 4</h1>
<div
id="data-normalization-and-identification-of-genes-driving-variability-cluster-formation"
class="section level2">
<h2>Data Normalization and Identification of Genes Driving Variability
(cluster formation)</h2>
<p>Normalize the matrix and create a plot of the genes that are the most
variable in the dataset.</p>
<p><strong>Glom_merged &lt;- NormalizeData(Glom_merged,
normalization.method = “LogNormalize”, scale.factor =
10000)</strong></p>
<p><strong>Glom_merged &lt;- FindVariableFeatures(Glom_merged,
selection.method = “vst”, nfeatures = 2000)</strong></p>
<p>This line identifies the 10 most highly variable genes and creates a
new object in the enviornment called “top10”.</p>
<p><strong>top10 &lt;- head(VariableFeatures(Glom_merged),
10)</strong></p>
<p>These lines creates a plot of the variable features within the
dataset and labels the top 10 genes to see which are driving the
clustering.</p>
<p><strong>plot1 &lt;- VariableFeaturePlot(Glom_merged)</strong></p>
<p><strong>plot2 &lt;- LabelPoints(plot = plot1, points = top10, repel =
TRUE)</strong></p>
<p><strong>plot2</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2015.png" /></p>
</div>
</div>
<div id="step-5" class="section level1">
<h1>Step 5</h1>
<div id="creation-and-visualization-of-principal-components"
class="section level2">
<h2>Creation and Visualization of Principal Components</h2>
<p>Calculate the principal components and then print the 5 genes that
drive the first 5 principal components. Can you get a sense for how the
cells might be clustering based off of these genes? Note: the
“ScaleData” and “RunPCA” steps can take very long with large
datasets</p>
<p><strong>all.genes &lt;- rownames(Glom_merged)</strong></p>
<p><strong>Glom_merged &lt;- ScaleData(Glom_merged, features =
all.genes)</strong></p>
<p><strong>Glom_merged &lt;- RunPCA(Glom_merged, features =
VariableFeatures(object = Glom_merged))</strong></p>
<p><strong>print(Glom_merged[[“pca”]], dims = 1:5, nfeatures =
5)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2016.png" /></p>
<p>Now we will visualize the genes driving the PCA a few ways. This line
visualizes the genes driving the first 2 PCA dimensions.</p>
<p><strong>VizDimLoadings(Glom_merged, dims = 1:2, reduction =
“pca”)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2017.png" /></p>
<p>This line visualizes the scRNAseq dataset as a PCA (hopefully you can
appreciate the poor clustering).</p>
<p><strong>DimPlot(Glom_merged, reduction = “pca”)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2018.png" /></p>
<p>This line creates a heatmap of the top 500 genes that drive the first
2 PCA dimensions.</p>
<p><strong>DimHeatmap(Glom_merged, dims = 1:2, cells = 500, balanced =
TRUE)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2019.png" /></p>
<p>This plot shows the variability and helps to decide how many PCA
dimensions to use when measuring the clusters. When the standard
deviation in the PC dimensions becomes small it no longer has a big
effect. In this case we will use 10 PC’s.</p>
<p><strong>ElbowPlot(Glom_merged)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2020.png" /></p>
</div>
</div>
<div id="step-6" class="section level1">
<h1>Step 6</h1>
<div id="calculation-and-creation-of-a-umap-visualization"
class="section level2">
<h2>Calculation and Creation of a UMAP visualization</h2>
<p>Determine the shape of the clusters and the number of populations in
our scRNASeq visualization. These lines control the resolution of the
clusters and the number of separate populations.</p>
<p>“FindNeighbors” changes the separation between clusters
“FindClusters” changes the number of differntial populations</p>
<p><strong>Glom_merged &lt;- FindNeighbors(Glom_merged, dims =
1:10)</strong></p>
<p><strong>Glom_merged &lt;- FindClusters(Glom_merged, resolution =
0.15)</strong></p>
<p><strong>Glom_merged &lt;- RunUMAP(Glom_merged, dims =
1:10)</strong></p>
<p>This line generates the UMAP visualization. Take a moment to reflect
on how do you think the clusters look.</p>
<p><strong>DimPlot(Glom_merged, reduction = “umap”)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2021.png" /></p>
<p>The first sanity/QC check for merged datasets is to see if cells are
clustering based on replicate. Since we have replicate info in the
metadata, we can check this using the “group.by” function.</p>
<p><strong>DimPlot(Glom_merged, reduction = “umap”, group.by =
“replicate”)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2022.png" /></p>
<p>The replicates are integrated nicely. It does not appear there is any
batch effect. If there is a batch effect, consider using the <a
href="https://satijalab.org/seurat/archive/v3.0/immune_alignment.html">“FindIntegrationAnchors”</a>
function, which we will not address in today’s lesson You can also
decrease the number of variable features in the “FindVariableFeatures”
function. Right now it is set to 2000, decreasing it will help deal with
batch effect, but will decrease the resolution of clustering.</p>
<p>We will next go through the cell identification features we covered
in the last lesson. First generation Heatmaps of differential genes
between clusters to identify cell types. This can take a VERY long time
with large datasets</p>
<p><strong>Glom_merged[[“RNA”]] &lt;-
JoinLayers(Glom_merged[[“RNA”]])</strong></p>
<p><strong>Glom.markers &lt;- FindAllMarkers(Glom_merged, only.pos =
TRUE, min.pct = 0.25, logfc.threshold = 0.25)</strong></p>
<p><strong>Glom.markers %&gt;% group_by(cluster) %&gt;% top_n(n = 5, wt
= avg_log2FC)</strong></p>
<p><strong>top5 &lt;- Glom.markers %&gt;% group_by(cluster) %&gt;%
top_n(5, avg_log2FC)</strong></p>
<p><strong>DoHeatmap(Glom_merged, features = top5$gene)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2023.png" /></p>
<p>This looks very similar to the heatmap we generated from the dataset
last lesson, let’s just double check that the identities are the
same.</p>
<p>Nphs2 = Podocyte Pecam1 = Endothelial Slc12a3 = Tubule Tagln =
Mural</p>
<p><strong>VlnPlot(Glom_merged, features = c(“Nphs2”, “Pecam1”,
“Slc12a3”, “Tagln”))</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2024.png" /></p>
<p>It looks like there are two different podocyte populations. We can
check out there differences between clusters 0 and 1 using the
“FindMarkers” function.</p>
<p><strong>cluster0.markers &lt;- FindMarkers(Glom_merged, ident.1 = 0,
ident.2 = 1, min.pct = 0.25)</strong></p>
<p><strong>head(cluster0.markers, n = 25)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2025.png" /></p>
<p>Using this information, we can visualize the differences using the
“Violin Plot” function.</p>
<p><strong>VlnPlot(Glom_merged, features = c(“Ctgf”))</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2026.png" /></p>
<p>We won’t spend a lot of time trying to identify different podocyte
populations, but you could look at more distinct populations, such as
endothelial (population 2) and epithelial (population 3) cells.</p>
<p><strong>cluster2.markers &lt;- FindMarkers(Glom_merged, ident.1 = 2,
ident.2 = 3, min.pct = 0.25)</strong></p>
<p><strong>head(cluster2.markers, n = 10)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2027.png" /></p>
</div>
</div>
<div id="step-7" class="section level1">
<h1>Step 7</h1>
<div id="re-name-the-clusters-by-cell-type-and-add-that-to-the-metadata"
class="section level2">
<h2>Re-name the clusters by cell-type, and add that to the metadata</h2>
<p><strong>new.cluster.ids &lt;- c(“Podo1”, “Podo2”, “Endo”, “Tubule”,
“Mural”)</strong></p>
<p><strong>names(new.cluster.ids) &lt;- levels(Glom_merged)</strong></p>
<p><strong>Glom_merged &lt;- RenameIdents(Glom_merged,
new.cluster.ids)</strong></p>
<p><strong>DimPlot(Glom_merged, reduction = “umap”, label = TRUE,
pt.size = 0.5, label.size = 7, repel = FALSE) + NoLegend()</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2028.png" /></p>
<p>Note that Podo1 and Podo2 are right next to each other, yet the Endo
population is split apart. I personally think this means the two
podocyte populations are not significant. The other Endo “cluster” may
represent doublets, but we will not spend much time addressing this. If
you want, you can spend time with the “FeaturePlot” function to
visualize what genes are expressed in the other Endo “Cluster”, then
adjust your “FindNeighbors” and “FindClusters” values a bit.</p>
<p><strong>FeaturePlot(Glom_merged, features = c(“Nphs2”))</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2029.png" /></p>
</div>
</div>
<div id="step-8" class="section level1">
<h1>Step 8</h1>
<div id="subclustering" class="section level2">
<h2>Subclustering</h2>
<p>Let’s do subclustering of the Tubule cells to demonstrate how one can
identify sub populations by isolating a certain cluster then
reclustering those cells. First, create a subset out of the tubule cells
using the “subset” function. We will make a new Seurat object called
“Glom_tubule”.</p>
<p><strong>Glom_tubule &lt;-subset(Glom_merged, idents =
c(“Tubule”))</strong></p>
<p><strong>Glom_tubule</strong></p>
<p>Glom_tuble is an object with 19183 features (genes) across 556
samples (cells).</p>
<p><img
src="images/lesson_2/roc_2_screenshot%2030.png" /></p>
<p>Next we will re-cluster the cells in Glom_tubule, starting with data
normalization.</p>
<p><strong>Glom_tubule &lt;- NormalizeData(Glom_tubule,
normalization.method = “LogNormalize”, scale.factor =
10000)</strong></p>
<p><strong>Glom_tubule &lt;- FindVariableFeatures(Glom_tubule,
selection.method = “vst”, nfeatures = 2000)</strong></p>
<p>Similarly, we will re-scale the data. Note that the row names are now
based on the row names of the new Seurat object.</p>
<p><strong>tubule.genes &lt;- rownames(Glom_tubule)</strong></p>
<p><strong>Glom_tubule &lt;- ScaleData(Glom_tubule, features =
tubule.genes)</strong></p>
<p><strong>Glom_tubule &lt;- RunPCA(Glom_tubule, features =
VariableFeatures(object = Glom_tubule))</strong></p>
<p><strong>print(Glom_tubule[[“pca”]], dims = 1:5, nfeatures =
5)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2031.png" /></p>
<p>Visualize the genes driving the first two PCAs with the following
code.</p>
<p><strong>VizDimLoadings(Glom_tubule, dims = 1:2, reduction =
“pca”)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2032.png" /></p>
<p>Also visualize the PCA.</p>
<p><strong>DimPlot(Glom_tubule, reduction = “pca”)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2033.png" /></p>
<p>Create a heatmap of the top 500 genes that drive the first 2 PCA
dimensions.</p>
<p><strong>DimHeatmap(Glom_tubule, dims = 1:2, cells = 500, balanced =
TRUE)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2034.png" /></p>
<p>Visualize the variance induced by each PC using the “Elbow Plot”
function.</p>
<p><strong>ElbowPlot(Glom_tubule)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2035.png" /></p>
<p>Now we can recluster the cells. Based on the elbow plot, we will set
the dims to 10 again.</p>
<p><strong>Glom_tubule &lt;- FindNeighbors(Glom_tubule, dims =
1:10)</strong></p>
<p><strong>Glom_tubule &lt;- FindClusters(Glom_tubule, resolution =
0.1)</strong></p>
<p><strong>Glom_tubule &lt;- RunUMAP(Glom_tubule, dims =
1:10)</strong></p>
<p>This line generates the UMAP visualization. How many clusters do you
see? Do they seem like they are real clusters?</p>
<p><strong>DimPlot(Glom_tubule, reduction = “umap”)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2036.png" /></p>
<p>Check to make sure the replicates aren’t driving the clustering.</p>
<p><strong>DimPlot(Glom_tubule, reduction = “umap”, group.by =
“replicate”)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2044.png" /></p>
<p>Let’s do a sanity check just to make sure we aren’t doing the same
cluster all over again. Do a violin plot of the same genes that defined
the original clusters.</p>
<p><strong>VlnPlot(Glom_tubule, features = c(“Nphs2”, “Pecam1”,
“Slc12a3”, “Tagln”))</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2037.png" /></p>
<p>It looks like there may be one population that contains
podocyte-tubule cell doublets, both otherwise looks pretty good. Now
let’s generate another heatmap of differential genes between clusters to
identify cell types.</p>
<p><strong>Tubule.markers &lt;- FindAllMarkers(Glom_tubule, only.pos =
TRUE, min.pct = 0.25, logfc.threshold = 0.25)</strong></p>
<p><strong>Tubule.markers %&gt;% group_by(cluster) %&gt;% top_n(n = 5,
wt = avg_log2FC)</strong></p>
<p><strong>top5_tubule &lt;- Tubule.markers %&gt;% group_by(cluster)
%&gt;% top_n(5, avg_log2FC)</strong></p>
<p><strong>DoHeatmap(Glom_tubule, features =
top5_tubule$gene)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2038.png" /></p>
<p>Pvalb, Egf, and Umod are all expressed by DCT.</p>
<p>Aqp2 is a marker of principal cells (PC) of the Collecting Duct.</p>
<p>Wt1 is expressed by podocytes.</p>
<p>Slc4a1 is expressed by intercalated cells (IC) of the collecting
duct.</p>
<p>Let’s check with a violin plot.</p>
<p><strong>VlnPlot(Glom_tubule, features = c(“Aqp2”, “Egf”, “Nphs2”,
“Atp6v1g3”))</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2039.png" /></p>
<p>Note the Egf is expressed in population 0 and 2, which suggests that
population 2 containes DCT/Podocyte doublets. Next, re-name the clusters
by cell-type, and add that to the metadata.</p>
<p><strong>new.cluster.ids.tubule &lt;- c(“DCT”, “PC”, “Doublet”,
“IC”)</strong></p>
<p><strong>names(new.cluster.ids.tubule) &lt;-
levels(Glom_tubule)</strong></p>
<p><strong>Glom_tubule &lt;- RenameIdents(Glom_tubule,
new.cluster.ids.tubule)</strong></p>
<p><strong>DimPlot(Glom_tubule, reduction = “umap”, label = TRUE,
pt.size = 0.5, label.size = 7, repel = FALSE) + NoLegend()</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2040.png" /></p>
<p>Now let’s get to know the clusters a little better. To determine how
many cells are in each cluster we can run the following code.</p>
<p><strong>table(Idents(Glom_tubule))</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2041.png" /></p>
<p>To determine the proportion of cells are in each cluster, run the
following code.</p>
<p><strong>prop.table(table(Idents(Glom_tubule)))</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2042.png" /></p>
<p>To determine the how does cluster membership vary by replicate, run
the following code.</p>
<p><strong>table(Idents(Glom_merged),
Glom_merged$replicate)</strong></p>
<p><strong>table(Idents(Glom_tubule),
Glom_tubule$replicate)</strong></p>
<p><img
src="images/lesson_2/roc_2_screenshot%2043.png" /></p>
<p>We can also save the “Tubule.markers” as a .csv so we can look over
it in excel, or use it for downstream analysis.</p>
<p><strong>write.csv(Tubule.markers,
“Glom/Tubule.markers.csv”)</strong></p>
<p>Lastly, we will save the “Glom_tubule” object so it can be loaded
later.</p>
</div>
</div>
<div id="step-9" class="section level1">
<h1>Step 9</h1>
<div id="save-seurat-object." class="section level2">
<h2>Save Seurat object.</h2>
<p>Will save into the folder that the Rscript file is located in.</p>
<p><strong>saveRDS(Glom_tubule, file =
“Glom/Glom_tutorial.rds”)</strong></p>
<p>To load a SeuratObject that was saved as an RDS, run the following
code.</p>
<p><strong>Glom_tubule &lt;-
readRDS(“Glom/Glom_tutorial.rds”)</strong></p>
</div>
</div>
<div id="extra-code-for-uploading-different-data-files-to-seuat"
class="section level1">
<h1>Extra code for uploading different data files to Seuat</h1>
<p>All three of these examples start with turning the data into a data
frame which Seurat can read, then creating a Seurat object out of the
data frame. Barcodes, features, and matrix files - common 10X output.
Make sure the three files (usually zipped) are in one folder. They must
be titled “barcodes.tsv.gz”, “features.tsv.gz”, and “matrix.mtx.gz”.</p>
<p>First upload the data using the “Read10X” function.</p>
<p><strong>Object.data &lt;- Read10X(data.dir = “Path to
file”)</strong></p>
<div id="h5-files" class="section level2">
<h2>H5 files</h2>
<p>To load H5 files, you will need to install the “hd5r” package.</p>
<p><strong>install.packages(“hdf5r”)</strong></p>
<p><strong>library(hdf5r)</strong></p>
<p>to upload the data, it must end in “.h5”.</p>
<p><strong>Object.data &lt;- Read10X_h5(“Path to .h5 file”)</strong></p>
</div>
<div id="dge-table" class="section level2">
<h2>DGE table</h2>
<p>To read DGE tables, use the the “read.table” function.The file must
be in “.txt” format. It will likely be in “.dge.txt” format.</p>
<p><strong>Object.data &lt;- read.table(file = “Path to .dge.txt”,
header = TRUE, row.names = 1, colClasses =c(“character”,
rep(“numeric”)))</strong></p>
<p>For all of these, make a Seurat object out of the data frame.</p>
<p><strong>Object_name &lt;- CreateSeuratObject(counts = Object.data,
min.cells = 3, min.genes = 200, project = “Project_name”)</strong></p>
</div>
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.4.2 (2024-10-31)
## Platform: aarch64-apple-darwin20
## Running under: macOS Sequoia 15.6.1
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## time zone: America/Los_Angeles
## tzcode source: internal
## 
## attached base packages:
## [1] grid      stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] kableExtra_1.4.0   openxlsx_4.2.8     viridis_0.6.5      viridisLite_0.4.2 
##  [5] gplots_3.2.0       reshape2_1.4.4     pheatmap_1.0.13    ggvenn_0.1.10     
##  [9] lubridate_1.9.4    forcats_1.0.0      stringr_1.5.1      purrr_1.1.0       
## [13] readr_2.1.5        tidyr_1.3.1        tibble_3.3.0       tidyverse_2.0.0   
## [17] devtools_2.4.5     usethis_3.1.0      here_1.0.1         ggpmisc_0.6.2     
## [21] ggpp_0.5.9         ggplot2_3.5.2      knitr_1.50         patchwork_1.3.1   
## [25] Seurat_5.3.0       SeuratObject_5.1.0 sp_2.2-0           dplyr_1.1.4       
## 
## loaded via a namespace (and not attached):
##   [1] RcppAnnoy_0.0.22       splines_4.4.2          later_1.4.2           
##   [4] bitops_1.0-9           polyclip_1.10-7        fastDummies_1.7.5     
##   [7] lifecycle_1.0.4        rprojroot_2.1.0        globals_0.18.0        
##  [10] lattice_0.22-7         MASS_7.3-65            magrittr_2.0.3        
##  [13] plotly_4.11.0          sass_0.4.10            rmarkdown_2.29        
##  [16] jquerylib_0.1.4        yaml_2.3.10            remotes_2.5.0         
##  [19] httpuv_1.6.16          sctransform_0.4.2      spam_2.11-1           
##  [22] zip_2.3.3              sessioninfo_1.2.3      pkgbuild_1.4.8        
##  [25] spatstat.sparse_3.1-0  reticulate_1.43.0      cowplot_1.2.0         
##  [28] pbapply_1.7-4          RColorBrewer_1.1-3     abind_1.4-8           
##  [31] pkgload_1.4.0          Rtsne_0.17             ggrepel_0.9.6         
##  [34] irlba_2.3.5.1          listenv_0.9.1          spatstat.utils_3.1-5  
##  [37] MatrixModels_0.5-4     goftest_1.2-3          RSpectra_0.16-2       
##  [40] spatstat.random_3.4-1  fitdistrplus_1.2-4     parallelly_1.45.1     
##  [43] svglite_2.2.1          codetools_0.2-20       xml2_1.3.8            
##  [46] tidyselect_1.2.1       farver_2.1.2           matrixStats_1.5.0     
##  [49] spatstat.explore_3.5-2 jsonlite_2.0.0         ellipsis_0.3.2        
##  [52] progressr_0.15.1       ggridges_0.5.6         survival_3.8-3        
##  [55] systemfonts_1.2.3      tools_4.4.2            ica_1.0-3             
##  [58] Rcpp_1.1.0             glue_1.8.0             gridExtra_2.3         
##  [61] xfun_0.52              withr_3.0.2            fastmap_1.2.0         
##  [64] SparseM_1.84-2         caTools_1.18.3         digest_0.6.37         
##  [67] timechange_0.3.0       R6_2.6.1               mime_0.13             
##  [70] textshaping_1.0.1      colorspace_2.1-1       scattermore_1.2       
##  [73] gtools_3.9.5           tensor_1.5.1           spatstat.data_3.1-6   
##  [76] generics_0.1.4         data.table_1.17.8      httr_1.4.7            
##  [79] htmlwidgets_1.6.4      uwot_0.2.3             pkgconfig_2.0.3       
##  [82] gtable_0.3.6           lmtest_0.9-40          htmltools_0.5.8.1     
##  [85] profvis_0.4.0          dotCall64_1.2          scales_1.4.0          
##  [88] png_0.1-8              spatstat.univar_3.1-4  rstudioapi_0.17.1     
##  [91] tzdb_0.5.0             nlme_3.1-168           cachem_1.1.0          
##  [94] zoo_1.8-14             KernSmooth_2.23-26     parallel_4.4.2        
##  [97] miniUI_0.1.2           pillar_1.11.0          vctrs_0.6.5           
## [100] RANN_2.6.2             urlchecker_1.0.1       promises_1.3.3        
## [103] xtable_1.8-4           cluster_2.1.8.1        evaluate_1.0.4        
## [106] cli_3.6.5              compiler_4.4.2         rlang_1.1.6           
## [109] future.apply_1.20.0    plyr_1.8.9             fs_1.6.6              
## [112] stringi_1.8.7          deldir_2.0-4           lazyeval_0.2.2        
## [115] spatstat.geom_3.5-0    quantreg_6.1           Matrix_1.7-3          
## [118] RcppHNSW_0.6.0         hms_1.1.3              future_1.67.0         
## [121] shiny_1.11.1           ROCR_1.0-11            igraph_2.1.4          
## [124] memoise_2.0.1          bslib_0.9.0            polynom_1.4-1</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4,h5",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
