<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Using Packages</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Learning Bioinformatics</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    About
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="about.html">About</a>
    </li>
    <li>
      <a href="contributors.html">Contributors</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Getting Started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="computer_requirements.html">Computer Requirements</a>
    </li>
    <li>
      <a href="downloading_R.html">Downloading R</a>
    </li>
    <li>
      <a href="using_packages.html">Using Packages</a>
    </li>
    <li>
      <a href="best_practices.html">Jonathan's Best Coding Practices</a>
    </li>
    <li>
      <a href="rules_of_coding.html">Jonathan's Rule of Coding</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Github Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="qupath_install.html">Understanding Github</a>
    </li>
    <li>
      <a href="github_tutorial.html">Github Tutorial</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    snRNAseq Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
<<<<<<< HEAD
      <a href="roc_1.html">ROC 1</a>
=======
      <a href="write_degs_excel.html">Write DEGs to Excel</a>
    </li>
    <li>
      <a href="coding_crumbs.html">SCORE: Coding Crumbs</a>
    </li>
    <li>
      <a href="recreating_figures.html">SCORE: Recreating Figures</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    QuPath Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="qupath_install.html">Installing QuPath</a>
    </li>
    <li>
      <a href="qupath_manual_annotation.html">Manual Cell Annotation</a>
    </li>
    <li>
      <a href="qupath_composite_classifications.html">Creating Composite Classifications</a>
    </li>
    <li>
      <a href="qupath_analyze_in_r.html">Analyzing QuPath data in R</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Problem Solving
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="write_degs_excel.html">Write DEGs to Excel</a>
>>>>>>> main
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Using Packages</h1>

</div>


<div id="intro-to-packages" class="section level1">
<h1>Intro to Packages</h1>
<p>Packages are collections of functions that help users perform
specific tasks. They are useful for visualization, statistical analysis,
data manipulation, reproducibility, and even bioinformatics itself. For
this example, you will learn how to install <strong>ggplot2</strong>
which is used for advanced plot creation as well as
<strong>dplyr</strong> which is used for cleaning and transforming
data.</p>
</div>
<div id="installing-packages" class="section level1">
<h1>Installing packages</h1>
<div id="step-1-download-the-package" class="section level2">
<h2>Step 1: Download the package</h2>
<div id="step-1a-download-one-package" class="section level3">
<h3>Step 1a: Download one package</h3>
<p>There are many ways to download a package. The easiest and most
straight forward way to do so is by using the
<strong>install.packages</strong> function. This is a base R function
which means you do not need to load any libraries to use it.</p>
<p>To download a package using <strong>install.packages</strong> you
simply need to type it or copy it into your R Script and run it. To run
a line of code on a Mac click on the number to the left of the line,
press Command + Enter. For Windows the procedure is the same but you
should press Control + Enter. In this example I am downloading the
package <strong>ggplot2</strong>.</p>
<p><img
src="Images/PG_Screenshot%201.png" /></p>
</div>
<div id="step-1b-download-multiple-packages" class="section level3">
<h3>Step 1b: Download multiple packages</h3>
<p>To run many lines of code press the number to the left of at the
start of the series of code you want to run then hold and drag to the
number to the left of the end of the series of code you want to run.
Alternatively you can click the Run button in the top right corner which
will run the entire R Script. In this example I am downloading the
packages <strong>ggplot2</strong> and <strong>dplyr</strong>.</p>
<p><img
src="Images/PG_Screenshot%202.png" /></p>
</div>
</div>
</div>
<div id="verfying-packages" class="section level1">
<h1>Verfying packages</h1>
<p>There are a number of steps to take to verify you have successfully
downloaded a package. The first marker that you are on the right path is
when you get this message in the R console.</p>
<p><img
src="Images/PG_Screenshot%203.png" /></p>
<p>Secondly, you will want to switch from the default
<strong>Plots</strong> page to the <strong>Packages</strong> page to
check that the packages has been downloaded. <strong>Packages</strong>
that have been installed but not loaded in the session will appear
unchecked in the <strong>Packages</strong> section as seen below.</p>
<p><img
src="Images/PG_Screenshot%204.png" /></p>
</div>
<div id="special-packages" class="section level1">
<h1>Special Packages</h1>
<p>Some packages you cannot access from R and will have to get from
outside sources, like GitHub. For example, the package
<strong>DoubletFinder</strong> is useful for predicting doublets in
scRNA-seq data which cleans the data and makes it easier to work with.
To download a package from GitHub you will access the users GitHub page
and copy the line to install.</p>
<p>remotes::install_github(‘chris-mcginnis-ucsf/DoubletFinder’)</p>
<p><img
src="Images/PG_Screenshot%205.png" /></p>
</div>
<div id="trouble-shooting" class="section level1">
<h1>Trouble Shooting</h1>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
